### DISCUSSION TITLE
FluentLayout vs FluentMainLayout

### DISCUSSION THREAD

**Discussion by mxmissile** [microsoft/fluentui-blazor#1881] [Q&A] [answered]
When would you use one over the other? And why?

**Comment by vnbaaij** [comment, marked as answer]
- `FluentLayout` just has `ChildContent` this means you are totally free on what you want to render inside of it. For example, yYou define if you want a header and what it will look like

- `FluentMainLayout` gives you parameters for header, sub-header, etc. If you provide content for those renderfragments they will be shown at set 'positions' in the layout. So it is more pre-defined on what it will look like. The `Header` will be rendered with the `FluentHeader`, etc

**Reply by Eagle3386** [reply]
If that's how it's supposed to work, shouldn't there be a `Footer` property added to `FluentMainLayout`, too?
Because right now, trying to do this, lets VS (or R#, can't tell for sure 😅) complete this as `<footer></footer>`, i.e., the HTML5 standard tag as the mentioned property is missing.

Thinking about this more, I also like to request a couple more features & improvements:
1. Either turn `FluentMainLayout.NavMenuTitle` into a `RenderFragment` or, better yet, refactor it into `NavMenuHeader`.
This allows for header customization by users if needed while continuing to support those who simply assign a `string`.
2. `NavMenuWidth` should've rather a responsive width (while maintaining proper left & right padding, that is) than being fixed to `200px` by default.
3. Resembling a "more pre-defined on what it will look like"-style as @vnbaaij called it, `FluentMainLayout` should offer `RenderFragment`s for app icon, notification & user menu as well - ultimately, it should contain sorta "wireframe" to get especially beginners up to speed as quickly as possible.
4. Speaking of simplicity, I'd like to ultimately suggest dropping the requirement for this:
```razor
<FluentMainLayout NavMenuTitle="Main menu">
  <Body>
    <FluentBodyContent>@Body</FluentBodyContent>
    <AppFooter /> @* Workaround until, e.g., <Footer> exists. *@
  </Body>
  <Header>
    <AppHeader @bind-IsNavMenuOpen="IsNavMenuOpen" />
  </Header>
  <NavMenuContent>
    <AppNavMenu @bind-IsNavMenuOpen="IsNavMenuOpen" />
  </NavMenuContent>
</FluentMainLayout>
```
and instead allow for either this:
```razor
<FluentMainLayout Body="@Body" Footer="AppFooter" Header="AppHeader" NavMenuContent="AppNavMenu"
                  NavMenuHeader="AppNavMenuHeader" @bind-IsNavMenuOpen="IsNavMenuOpen" />
```
or at least this:
```razor
<FluentMainLayout @bind-IsNavMenuOpen="IsNavMenuOpen">
  @Body @* Should leverage auto-assignment via FluentMainLayout.ChildContent instead. *@
  <Header>
    <AppHeader />
  </Header>
  <NavMenu>
    <Header> @* Optional *@
      <AppNavMenuHeader />
    </Header>
    <AppNavMenu /> @* Leverages auto-assignment via NavMenu.ChildContent, too. *@
  </NavMenu>
  <Footer>
    <AppFooter />
  </Footer>
</FluentMainLayout>
```
Note 1: `@Body` should be set as `ChildContent` of an _automatically_ inserted `FluentBodyContent`.
Note 2: The first proposal's attributes/properties, `Footer`, `Header` & `NavMenu[Content/Header]`, should support assignment of only a component's name & handling "the rest" automatically.
Note 3: In any case, the `@bind` for `IsNavMenuOpen` should happen on the root, i.e., `FluentMainLayout`.
That way, it's
- always automatically handled by `FluentMainLayout` itself,
- pushed down to `NavMenu[Content/Header]` & potentially `NotificationPanel` as well as `ProfileMenu` (e.g., for a quick toggle there), if need be
- even allows for "subscription" within `MainLayout.razor` for custom user scenarios.

Again, as stated earlier, the concept of `FluentLayout` allowing complete control while `FluentMainLayout` offering a more pre-defined wireframe is just awesome & I _really_ love to see this taken further - heck, I love it so much, if granted, I'd provide the code for it all by myself! 😍

What do you guys think? 😎

**Reply by vnbaaij** [reply]
Sounds good. Looking forward to your PR's!

**Reply by Eagle3386** [reply]
@vnbaaij Would you want me to go for 1st ("attribute syntax"), rather the 2nd ("child-node simplification") or should I do both? 🤔

**Reply by vnbaaij** [reply]
I have a preference for the second one ("child-node simplification"), so maybe start there?

**Reply by Eagle3386** [reply]
Sure, your wish is my command! 😉😅
Just give me a couple of days, please. Real-life is currently quite "eager" at soaking it all in, every single free minute that might appear.. 🙈
